Fail self-recursive deferred given impl

som-snytt opened the pull request:
At typer, fail if implementing a deferred given picks the implementing member, which can happen in the companion to the type class.

The recursion is not detected by `CheckLoopingImplicits`, where a warning might have sufficed, because the RHS gets the symbol of the deferred member and not its implementation.

It behooves typer to refuse to infer the member in this case.

The feature is intended to reduce boilerplate where the implicit value is found in lexical scope, so that not too much magic is demanded.

Fixes #22589

hamzaremmal reviewed (changes requested):
Can you also squash everything in a single commit

som-snytt commented:
It's harder for me to re-review in squashed form. It looks like a minor refactor to reuse the reporting and also make the `if/else` easier to read. The `resolvedHere` bit is new.

som-snytt commented:
rebased

som-snytt commented:
Refactored `responsibleForImports` fix to a def for scoping; that conflicted with the previous refactor to reduce the big else branch.

Also refactored the previous warning to `rhs.tap(warn)`, also for scoping. The near pun is `tap(warn)` and `tapeworm`. The usual pun is `warn` and `worn`, which also applies.

That was three months ago.

Gedochao commented:
(rebased)

noti0na1 reviewed (approved):
LGTM
